scripts/_helpers.py

```diff
@@ -150,3 +152,81 @@ def dlProgress(count, blockSize, totalSize):
 
     urllib.request.urlretrieve(url, file, reporthook=dlProgress)
 
+
+def mock_snakemake(rulename, **wildcards):
+    """
+    This function is expected to be executed in the 'scripts'-directory within
+    the snakemake project. It returns a snakemake.script.Snakemake object,
+    based on the Snakefile of the upper directory, such that includes
+    mostly all necessary information (input, output, log etc.).
+
+    If a rule has wildcards, you have to specify them in **wildcards.
+
+    Parameters
+    ----------
+    rulename: str
+        name of the rule for which the snakemake object should be generated
+    **wildcards:
+        keyword arguments fixing the wildcards. Only necessary if wildcards are
+        needed.
+    """
+    import snakemake as sm
+    import os
+    from pypsa.descriptors import Dict
+    from os.path import abspath
+
+    base_dir = Path(__file__).parent.parent
+    old_wd = os.getcwd()
+    os.chdir(base_dir)
+    for p in sm.SNAKEFILE_CHOICES:
+        if os.path.exists(p):
+            snakefile = p
+            break
+    workflow = sm.Workflow(snakefile)
+    workflow.include(snakefile)
+    rule = workflow.get_rule(rulename)
+    wc = Dict(wildcards)
+
+    # make the input files accessable by taking the absolut paths
+    def make_io_accessable(smfiles):
+        if not smfiles:
+            return []
+        # for mildy hacky input functions (like make_summary):
+        if len(smfiles) == 1 and callable(smfiles[0]):
+            new_input_files = smfiles[0](wc)
+            # overwrite smfiles with extracted files from function
+            smfiles = sm.io.InputFiles()
+            for index, p in enumerate(new_input_files):
+                smfiles.insert(index, p)
+        # now iterate over each item and make path an absolut path
+        files = getattr(sm.io, type(smfiles))()
+        for index, (key, p) in enumerate(smfiles.allitems()):
+            # case that item is a function
+            if callable(p):
+                p = p(wc)
+            if isinstance(p, list):
+                if len(p) == 1:
+                    p = p[0]
+            if isinstance(p, str):
+                files.insert(index, sm.io.apply_wildcards(abspath(p), wc))
+            else:
+                files.insert(index, p)
+            if key is not None:
+                files.add_name(key)
+        return files
```

Comment:
Seeing now, the difficulty with which you are trying to patch all available paths to work with whatever path one started from. I fear that this code is prone to break soon, as you noticed with the log file fallbacks for example.

I think it's better to take a radically different path and make clear, that if you want to run a script from a python interpreter one has to start the interpreter from the repository root, where the Snakefile is located, and simplify all of this, drastically.

Reply:
@coroa you are right, about the simplification. However, I would like to keep it like that for now, as I have barely time this and next week, and going through the snakemake code takes a lot of time. I'll set an (self-referenced) issue for that as soon as the PR is through.

Reply:
@FabianHofmann I think you misunderstood. I would propose to throw the towel and reduce everything down to the following and be done with it. You're trying to crawl down a rabbit hole.

```python
def mock_snakemake(rulename, **wildcards):
    """
    This function is expected to be executed in the 'scripts'-directory within
    the snakemake project. It returns a snakemake.script.Snakemake object,
    based on the Snakefile of the upper directory, such that includes
    mostly all necessary information (input, output, log etc.).

    If a rule has wildcards, you have to specify them in **wildcards.

    Parameters
    ----------
    rulename: str
        name of the rule for which the snakemake object should be generated
    **wildcards:
        keyword arguments fixing the wildcards. Only necessary if wildcards are
        needed.
    """
    import snakemake as sm
    import os
    from pypsa.descriptors import Dict
    from os.path import abspath

    for p in sm.SNAKEFILE_CHOICES:
        if os.path.exists(p):
            snakefile = p
            break
    workflow = sm.Workflow(snakefile)
    workflow.include(snakefile)
    rule = workflow.get_rule(rulename)
    wc = Dict(wildcards)

    # create log and output dir if not existent
    for path in chain(rule.log.values(), rule.output.values()):
        Path(path).parent.mkdir(parents=True, exist_ok=True)

    snakemake = sm.script.Snakemake(input=rule.input, output=rule.output,
                                    params=rule.params, wildcards=wc, threads=None,
                                    resources=rule.resources, log=rule.log,
                                    config=workflow.config, rulename=rule.name,
                                    bench_iteration=None)

    return snakemake
```

Reply:
But this is not working, I wished it was as easy. The problem is that the wildcards are not applied here (which made the whole thing a bit tricky). 
I am sure there is an easier way to do this than I did, but it requires going through the snakemake code and look how they create the `snakemake` object.

Reply:
initializing the `snakemake` class is literately just passing objects (see [here](https://github.com/snakemake/snakemake/blob/5bb0ed743c8850532a8d4c485105fefc9cf33bc5/snakemake/script.py#L167))

Reply:
I think I have it:

```python
def mock_snakemake(rulename, **wildcards):
    """
    This function is expected to be executed from the 'scripts'-directory of the
    the snakemake project. It returns a snakemake.script.Snakemake object,
    based on the Snakefile.

    If a rule has wildcards, you have to specify them in **wildcards.

    Parameters
    ----------
    rulename: str
        name of the rule for which the snakemake object should be generated
    **wildcards:
        keyword arguments fixing the wildcards. Only necessary if wildcards are
        needed.
    """
    import snakemake as sm
    import os
    from pypsa.descriptors import Dict
    from snakemake.script import Snakemake
    import logging

    base_dir = Path(__file__).parent.joinpath('..').absolute()
    if Path(os.getcwd()).absolute() != base_dir:
        logging.info(f'Changing directory to repository root {base_dir}')
        os.chdir(base_dir)

    for p in sm.SNAKEFILE_CHOICES:
        if os.path.exists(p):
            snakefile = p
            break
    workflow = sm.Workflow(snakefile)
    workflow.include(snakefile)
    workflow.global_resources = {}
    rule = workflow.get_rule(rulename)
    dag = sm.dag.DAG(workflow, rules=[rule])
    wc = Dict(wildcards)
    job = sm.jobs.Job(rule, dag, wc)
    snakemake = Snakemake(job.input, job.output, job.params, job.wildcards,
                          job.threads, job.resources, job.log,
                          job.dag.workflow.config, job.rule.name, None,)

    # create log and output dir if not existent
    for path in list(rule.log) + list(rule.output):
        Path(path).parent.mkdir(parents=True, exist_ok=True)
    return snakemake
```
does the thing. I know there is a change of the directory in the function itself which is nasty. But it has to be done there as all the relative imports within the scripts would not work otherwise, like (`from _helpers import configure_logging` or `from add_electricity import ...`)

Reply:
Ok, so using this function will (in some cases) change my current working directory without changing it back? I am curious to see how this holds up in the real world.

Minor correction of my previous statement, based on @coroa 's comment:
It should be `.resolve()`, not `.absolute()` (which is neither tested nor officially documented). So this
```
base_dir = Path(__file__).parent.joinpath('..').absolute()
    if Path(os.getcwd()).absolute() != base_dir:
```
should really be this:
```
base_dir = Path(__file__).resolve().parents[1]
    if Path.cwd().resolve() != base_dir:
```

Reply:
great, thanks for the suggestion. I pushed the new function, and see where it gets... It really takes  effort to understand `Snakemake` properly. It works well for all scripts, only exception is that the logging sometime behaves weird. Thankful if someone can test it.

Reply:
Instead of changing the working directory, it would also be fine to fail with an error from which directory one has to run the code!

Reply:
true, but then (as mentioned above) the relative module imports won't work...

Reply:
Ah sorry, I didn't get that. Why is that? Can the relative imports be written differently?

I don't immediately see how, though.

Reply:
I resolve this conversation, since it's outdated. Just as a side note:
I tried a lot with importing the modules differently, one approach was to import them as 
```python 
from scripts._helpers import ...
```
But this did not work for the normal Snakemake workflow as the repo root, `pypsa-eur`, is not in the paths (sys.path). Trying to add `pypsa-eur` to sys.path in the beginning of the Snakefile does not affect the execution of a rule, as a different thread process is opened for this.